File: app/src/main/java/de/dizcza/fundu_moto_joystick/fragment/LogsFragment.java
package de.dizcza.fundu_moto_joystick.fragment;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import de.dizcza.fundu_moto_joystick.R;
import de.dizcza.fundu_moto_joystick.util.Constants;

public class LogsFragment extends Fragment {

    private final LogsView mStatusView = new LogsView();
    private final LogsView mSentView = new LogsView();
    private final LogsView mReceivedView = new LogsView();
    private JoystickFragment mJoystickFragment;

    private static class LogsView {
        TextView view;
        final StringBuilder logs = new StringBuilder();

        private void setView(TextView view) {
            this.view = view;
        }
    }

    public LogsFragment() {
    }

    public void setJoystickFragment(JoystickFragment joystickFragment) {
        mJoystickFragment = joystickFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        setRetainInstance(true);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.logs_fragment, container, false);
        mStatusView.setView(view.findViewById(R.id.logs_status));
        mSentView.setView(view.findViewById(R.id.logs_sent));
        mReceivedView.setView(view.findViewById(R.id.logs_received));
        final EditText commandView = view.findViewById(R.id.send_text);
        final ImageButton sendCmdButton = view.findViewById(R.id.send_btn);
        sendCmdButton.setOnClickListener(v -> {
            final String command = commandView.getText().toString();
            if (mJoystickFragment.send(command.getBytes())) {
                mSentView.view.append(command + Constants.NEW_LINE);
            }
        });
        return view;
    }

    private Iterable<LogsView> getViews() {
        return Arrays.asList(mStatusView, mSentView, mReceivedView);
    }

    @Override
    public void onResume() {
        super.onResume();
        for (LogsView logsView : getViews()) {
            String text = logsView.logs.toString();
            List<String> textLines = Arrays.asList(text.split(Constants.NEW_LINE));
            Collections.reverse(textLines);
            String textReversed = text.join(Constants.NEW_LINE, textLines);
            logsView.view.setText(textReversed);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_logs, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        final int idSelected = item.getItemId();
        if (idSelected == R.id.clear) {
            for (LogsView logsView : getViews()) {
                logsView.view.setText("");
                logsView.logs.setLength(0);
            }
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    public void appendStatus(String text) {
        mStatusView.logs.append(text);
    }


    public void appendSent(String text) {
        mSentView.logs.append(text);
    }


    public void appendReceived(String text) {
        mReceivedView.logs.append(text);
    }

}

File: app/src/main/java/de/dizcza/fundu_moto_joystick/fragment/LogsFragment.kt
package de.dizcza.fundu_moto_joystick.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.TextView
import androidx.fragment.app.Fragment
import de.dizcza.fundu_moto_joystick.R
import de.dizcza.fundu_moto_joystick.util.Constants
import java.util.Arrays
import java.util.Collections

class LogsFragment : Fragment() {
    private val mStatusView = LogsView()
    private val mSentView = LogsView()
    private val mReceivedView = LogsView()
    private var mJoystickFragment: JoystickFragment? = null

    private class LogsView {
        var view: TextView? = null
        val logs = StringBuilder()
    }

    fun setJoystickFragment(joystickFragment: JoystickFragment?) {
        mJoystickFragment = joystickFragment
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        retainInstance = true
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.logs_fragment, container, false)
        mStatusView.view = view.findViewById(R.id.logs_status)
        mSentView.view = view.findViewById(R.id.logs_sent)
        mReceivedView.view = view.findViewById(R.id.logs_received)
        val commandView = view.findViewById<EditText>(R.id.send_text)
        val sendCmdButton = view.findViewById<ImageButton>(R.id.send_btn)
        sendCmdButton.setOnClickListener { v: View? ->
            val command = commandView.text.toString()
            if (mJoystickFragment!!.send(command.toByteArray())) {
                mSentView.view!!.append(command + Constants.NEW_LINE)
            }
        }
        return view
    }

    private val views: Iterable<LogsView>
        private get() = Arrays.asList(mStatusView, mSentView, mReceivedView)

    override fun onResume() {
        super.onResume()
        for (logsView in views) {
            val text = logsView.logs.toString()
            val textLines = Arrays.asList(
                *text.split(Constants.NEW_LINE.toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray())
            Collections.reverse(textLines)
            val textReversed = java.lang.String.join(Constants.NEW_LINE, textLines)
            logsView.view!!.text = textReversed
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_logs, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val idSelected = item.itemId
        return if (idSelected == R.id.clear) {
            for (logsView in views) {
                logsView.view!!.text = ""
                logsView.logs.setLength(0)
            }
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    fun appendStatus(text: String?) {
        mStatusView.logs.append(text)
    }

    fun appendSent(text: String?) {
        mSentView.logs.append(text)
    }

    fun appendReceived(text: String?) {
        mReceivedView.logs.append(text)
    }
}
